### ISSUE TITLE
Obsolete ListMapper

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#3381]
`ListMapper.Property` throws NullReferenceException if you pass a collection with a null element to `new ListMapper`. I was going to fix it to throw a proper ArgumentException, then I realized that it didn't seem very useful and in fact is not used (or tested) at all.

Since it's connected to the AssertionHelper class which is already obsoleted, could we delete it? Or should we mark it obsolete instead? We shouldn't delete it if we think someone out there is using it, because they might have a binary dependency on it. I'm having a hard time picturing someone using it, given what it does.

One benefit to deleting it is that intellisense won't suggest it if you type `List` without angle brackets and press Ctrl+. to add a using directive.

```diff
namespace NUnit.Framework
{
    [Obsolete("The AssertionHelper class has been deprecated and will be removed in a future "
        + "release. Please consider using the NUnit.StaticExpect NuGet package instead.")]
    public class AssertionHelper
    {
-        public ListMapper Map(ICollection original);
    }

-   public class List
-   {
-       public static ListMapper Map(ICollection actual);
-   }

-   public class ListMapper
-   {
-       public ListMapper(ICollection original);

-       public ICollection Property(string name);
-   }
}
```

**Comment by CharliePoole** [comment]
Why do you say that `List` / `ListMapper` is connected to `AssertionHelper`?

`List.Map` was created as a way to do things that we do nowadays using `System.Linq`. 

Of course, `AssertionHelper` uses `ListMapper` allowing the user to avoid the static syntax, but it also allows the user to avoid `Assert`, which we presumably are not planning to obsolete. :smiling_imp: 

That's not to say there could be reasons to obsolete it. The main one might be if you don't want to support it in NUnit 4, which seems reasonable. But currently we have no way to know who is using it although I imagine they would turn up if you removed it.

**Comment by jnm2** [comment]
> Why do you say that List / ListMapper is connected to AssertionHelper?

It's the only thing in our codebase that references `List` and `List` is the only thing that references `ListMapper`. I guess that really doesn't mean much.

How would this have been used- something like this?


```cs
Assert.That(List.Map(items).Property("Foo"), Is.EqualTo(new[] { 1, 2, 3 }));
```

Would we recommend this?

```cs
Assert.That(items.Select(i => i.Foo), Is.EqualTo(new[] { 1, 2, 3 }));
```

Should we think about adding constraint syntax to do the same thing and provide a richer message?

**Comment by CharliePoole** [comment]
Exactly... it was created before .NET 3.5 came out.

I always wanted to have a fluent syntax to do that. List.Map was a "temporary" place holder, which became unnecessary as soon as Linq was available.

The syntax I was thinking of was something like

`Assert.That(items, Has.Each.Property("foo").EqualTo(new[] { 1, 2, 3}));`

Not sure it's really any better than using `Select`, however.

**Comment by jnm2** [comment]
All the syntaxes I can think of sound ambiguous, e.g. "Each item has a property `Foo` which contains the same three values," so you're probably right. One benefit of constraint model versus .Select would be that we could build a richer failure message, but there's always a line somewhere.

Submitted https://github.com/nunit/nunit/pull/3382.

**Comment by mikkelbu** [comment]
The documentation has the following examples on usage of ListMapper - https://github.com/nunit/docs/wiki/ListMapper
```csharp
string[] strings = new string[] { "a", "ab", "abc" };
int[] lengths = new int[] { 1, 2, 3 };

Assert.That(List.Map(strings).Property("Length"), 
       Is.EqualTo(lengths));
	   
Assert.That(new ListMapper(strings).Property("Length"),
       Is.EqualTo(lengths));
```
I've searched Github to figure out how much this is used, but most usages of the ListMapper are from copies of the source code of Nunit 2.X or from forks of `nunit-csharp-samples`. I've only found a handful of proper usages. I'm sure that there are more, but it is not a very used feature. So I'm also in favor of removing/obsoleting it.
